src/autoprops.rs

```diff
@@ -0,0 +1,260 @@
+        } = &self.item_fn;
+
+        let fn_name = &sig.ident;
+        let (_impl_generics, type_generics, where_clause) = sig.generics.split_for_impl();
+        let inputs = if self.needs_a_properties_struct() {
+            // NOTE: function components currently don't accept receivers, we're just passing the
+            //       information to the next macro to fail and give its own error message
+            let receivers = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Receiver(receiver) => Some(receiver),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let args = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Typed(syn::PatType { pat, .. }) => Some(quote! { #pat, }),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let phantom = sig.generics.type_params().next().is_some().then(|| {
+                quote! {
+                    phantom: _,
+                }
+            });
+            quote! {
+                #(#receivers,)* #properties_name {
+                    #(#args)*
+                    #phantom
+                }: &#properties_name #type_generics
+            }
+        } else {
+            quote! {}
+        };
+        let clones = sig
+            .inputs
+            .iter()
+            .filter_map(|arg| match arg {
+                syn::FnArg::Typed(syn::PatType { pat, ty, .. })
+                    if !matches!(**ty, syn::Type::Reference(_)) =>
+                {
+                    Some(quote! {
+                        let #pat = ::yew::html::ImplicitClone::implicit_clone(#pat);
+                    })
+                }
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        let output = &sig.output;
+        let impl_generics_with_defaults = &sig.generics;
+
+        quote! {
+            #(#attrs)*
+            #vis fn #fn_name #impl_generics_with_defaults (#inputs) #output #where_clause {
+                #(#clones)*
+                #block
+            }
+        }
+    }
+
+    fn print_properties_struct(&self) -> proc_macro2::TokenStream {
+        let properties_name = &self.properties_name;
+        let syn::ItemFn { vis, sig, .. } = &self.item_fn;
+
+        if !self.needs_a_properties_struct() {
+            return quote! {};
+        }
+
+        let (impl_generics, type_generics, where_clause) = sig.generics.split_for_impl();
+        let fields = sig
+            .inputs
+            .iter()
+            .filter_map(|arg| match arg {
+                syn::FnArg::Typed(syn::PatType { attrs, pat, ty, .. }) => match ty.as_ref() {
+                    syn::Type::Reference(syn::TypeReference { elem, .. }) => {
+                        Some(quote! { #(#attrs)* #vis #pat: #elem, })
```

Comment:
It is debatable of whether attributes of function arguments should be applied to the function itself or to properties struct definition. `#[props_...]` should be transferred to struct definition, however, I am not sure about other attributes.

All of the following order of attribute macros would result in different code generation behaviour to `#[foreign_trait]` if `#[foreign_trait]` will process attributes based on `#[foreign_trait_attr]`.

```rust
#[foreign_trait]
#[autoprops]
#[function_component]
fn Comp(#[foreign_trait_attr] arg: Arg) -> Html {}
```

```rust
#[autoprops]
#[foreign_trait]
#[function_component]
fn Comp(#[foreign_trait_attr] arg: Arg) -> Html {}
```

```rust
#[autoprops]
#[function_component]
#[foreign_trait]
fn Comp(#[foreign_trait_attr] arg: Arg) -> Html {}
```

Reply:
Yes, `#[foreign_trait]` (should've been called `#[foreign_proc_macro]`?) will make it be handled differently in those 3 cases, and this is up to the user to remember. You would not swap lines of function's code without noticing that it may change its logic. As I was writing this comment, I came to a realisation that it does not really make sense to put anything other than `#[function_component]` inside `#[autoprops]`, and any other `#[foreign_trait]` should transform the fn before `#[autoprops]` kicks in, where `#[autoprops]` would not even realise that there is any work of the external proc macros.

Reply:
Maybe ban any `#[foreign_trait_attr]`s?

Reply:
I think it's a bit too hypothetical. It might be best to wait that an actual come up?

Reply:
@cecton
The
> I am OK with leaving the implementation as-is

probably means its ok with us i guess, and can be marked as resolved, but
> Maybe ban any `#[foreign_trait_attr]`s inside `#[autoprops]`?

kinda makes sense since the autoprops fn arguments dont behave like any other fn arguments (since they get transformed into struct, destructure, etc, anything but fn arguments as-is), if you are still not sure then we can leave it as it already is done

Reply:
I'm not sure because if someone makes another proc macro that can be used in the composition here, like:

```
#[foreign_trait]
#[autoprops]
#[function_component]
fn Comp(#[foreign_trait_attr] arg: Arg) -> Html {}
```

It could be that the foreign trait attr will be properly located. I think it's best to revisit this in the future and avoid being too restrictive now. When you are too restrictive you block a path and changing path requires a breaking change. While if you don't commit yet, you can still leave this as "undefined" and pick a direction later on without making a breaking change

Reply:
@cecton i already mentioned this for your example

> `#[foreign_trait]` should transform the fn before `#[autoprops]` kicks in, where `#[autoprops]` would not even realise that there is any work of the external proc macros

Reply:
ok I think there is miscommunication here. What do I need to do exactly? ^_^

Reply:
The idea is that foreign attrs (everything except `#[prop_or...]`, which are intentional - autoprops's fn arg attr `#[prop_or...]` turns into yew's struct field attr `#[prop_or...]`) that were applied to fn args now get applied to struct fields in your generated code, which sounds wrong. The idea is to ban anything but the `#[prop_or(...)]` because there arent any fn args like there were before autoprops transformation (e.g. before transformation `a: ..., b: ...`, after transformation `Props { pub a: ..., pub b: ... }`, fn args and struct fields are completely different things, so none fn arg attrs would work on struct field attrs, especially since we cannot know what should be a `#[derive(ForeignTrait)]`, but we know that `#[prop_or...]` works for `#[derive(Properties)]`).

Reason it works in your example is that if you put foreign attr before autoprops, it should strip those fn arg attrs out, so essentially autoprops only sees

```rust
#[autoprops]
#[function_component]
fn Comp(arg: Arg) -> Html {}
```

Reply:
And leaving things as is for now and see how it goes is not acceptable? We have to commit into banning these other attrs **now**? We can't keep this as an open question for later?

And what should I do? Raise an error or strip them silently? What if the user use `#[cfg(...)]`, I assume I still need to keep those?

Reply:
> What if the user use `#[cfg(...)]`

Should this be handled by autoprops too (copy to both destructure, struct definition, etc)? This asks for a separate PR/Issue open

> Raise an error or strip them silently?

Error, stripping them silently will be very unwanted

> And leaving things as is for now and see how it goes is not acceptable?

Arguable tbf, I do think it sounds wrong to put a fn arg attr to struct field without telling anything about it to the struct, I think it will error by itself because of this, so it might be fine, but the error would make not a lot of sense to the user, which is why I think its a good idea to fix it before making the change public, but if you are still not sure:

> We can't keep this as an open question for later?

As mentioned,

> I am OK with leaving the implementation as-is

> probably means its ok with us i guess, and can be marked as resolved

Reply:
Ok I made an issue because clearly it's complicated https://github.com/yewstack/yew-autoprops/issues/11